include/deal.II/fe/fe_values_coupling.h

```diff
@@ -0,0 +1,557 @@
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_fe_values_coupling_h
+#define dealii_fe_values_coupling_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/algorithms/general_data_storage.h>
+
+#include <deal.II/base/subscriptor.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/fe/fe_values_views.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+// Forward declaration
+#ifndef DOXYGEN
+template <int dim, int spacedim>
+class FEValuesBase;
+#endif
+
+namespace FEValuesViews
+{
+  /**
+   * A class that provides a renumbered view to a given FEValuesViews object.
+   *
+   * In general, the order of the degrees of freedom and quadrature points
+   * follows the one of the FEValues object. However, in some cases, it is
+   * convenient to group together degrees of freedom and quadrature points in a
+   * different order. This class provides a view to a given FEValuesViews
+   * object, where the degrees of freedom and quadrature points are renumbered
+   * according to the given renumbering vectors.
+   */
+  template <typename ViewType>
+  class ReorderedView : public Subscriptor
+  {
+  public:
+    /**
+     * An alias for the data type of values of the underlying view.
+     */
+    using value_type = typename ViewType::value_type;
+
+    /**
+     * An alias for the data type of gradients of the underlying view.
+     */
+    using gradient_type = typename ViewType::gradient_type;
+
+    /**
+     * An alias for the data type of the product of a @p Number and the values
+     * of the underlying view type.
+     */
+    template <typename Number>
+    using solution_value_type = typename ProductType<Number, value_type>::type;
+
+    /**
+     * An alias for the data type of the product of a @p Number and the gradients
+     * of the underlying view type.
+     */
+    template <typename Number>
+    using solution_gradient_type =
+      typename ProductType<Number, gradient_type>::type;
+
+    /**
+     * Construct a new ReorderedView object.
+     *
+     * The @p dof_renumbering vector is used to renumber the degrees of freedom,
+     * while the @p quadrature_renumbering vector is used to renumber the
+     * quadrature points. An empty renumbering vector simply means that no
+     * renumbering is performed.
+     *
+     * @note The renumbering vectors may *not* match the dimensions of the
+     * underlying view, i.e., you could use this class to only run over half of
+     * the degrees of freedom, and integrating only over half of the current
+     * cell, if you wish to do so. The important part is that every entry of
+     * each renumbering vector is a legal value for the underlying view object.
+     * We allow the dof renumbering vector to contain
+     * `numbers::invalid_unsigned_int` values, which are ignored, and produce
+     * zero values, gradients, hessians, etc. for the corresponding shape
+     * function.
```

Comment:
Why is this useful? If so, why not just give a shorter vector?

Separately, can you give a larger vector? One where some indices appears more than once?

Reply:
The idea is that I want to be able to combine two different FEValues objects, and gather together all dof indices. So yes, the typical use case is when the renumbering vector is longer, and some of the indices of the renumbering vector refer to the *other* fevalues, while some refer to this one, and when they rerfer to the other this one must be zero (pretty much like what we do for the components extraction).